jenkins/bootstrap.py

```diff
@@ -0,0 +1,438 @@
+    try:
+        cache = json.loads(Subprocess(cmd, output=True))
+        if not isinstance(cache, list):
+            raise ValueError(cache)
+    except (subprocess.CalledProcessError, ValueError):
+        cache = []
+    cache.append({
+        'version': version,
+        'buildnumber': build,
+        'passed': bool(passed),
+        'result': 'SUCCESS' if passed else 'FAILURE',
+    })
+    cache = cache[-200:]
+    gsutil.UploadJson(path, cache)
+
+
+
+def Finish(gsutil, paths, success, artifacts, build, version):
+    if os.path.isdir(artifacts):
+        UploadArtifacts(paths.artifacts, artifacts)
+
+    # Upload build-log.txt
+
+    # Upload the latest build for the job
+    for path in {paths.latest, paths.build_latest}:
+        gsutil.UploadText(path, str(build), cached=False)
+
+    # Upload a link to the build path in the directory
+    if paths.build_link:
+        gsutil.UploadText(paths.build_link, paths.build_path)
+
+    AppendBuild(gsutil, paths.result_cache, build, version, success)
+    if paths.build_result_cache:
+        AppendBuild(gsutil, paths.build_result_cache, build, version, success)
+
+    # update_job_result_cache ${JENKINS_BUILD_FINISHED}
+
+    data = {
+        'timestamp': time.time(),
+        'result': 'SUCCESS' if success else 'FAILURE',
+        'passed': bool(success),
+    }
+    # TODO(rmmh): update tooling to expect metadata in finished.json
+    metadata = os.path.join(paths.artifacts, 'metadata.json')
+    if os.path.isfile(metadata):
+        try:
+            with open(metadata) as fp:
+                val = json.loads(fp.read())
+        except (IOError, ValueError):
+            val = None
+        if val and isinstance(val, dict):
+            data['metadata'] = val
+    gsutil.UploadJson(paths.finished, data)
+
+
+
+
+def TestInfra(*paths):
+    """Return path relative to root of test-infra repo."""
+    return os.path.join(ORIG_CWD, os.path.dirname(__file__), '..', *paths)
+
+
+def Node():
+    if 'NODE_NAME' not in os.environ:
+        os.environ['NODE_NAME'] = ''.join(socket.gethostname().split('.')[:1])
+    return os.environ['NODE_NAME']
+
+
+def Version():
+    version_file = 'version'
+    if os.path.isfile(version_file):
+        # TODO(fejta): is this code path ever used?
+        with open(version_file) as fp:
+            return fp.read().strip()
+
+    version_script = 'hack/lib/version.sh'
+    if os.path.isfile(version_script):
+        cmd = [
+            'bash', '-c', (
+"""
+set -o errexit
+set -o nounset
+export KUBE_ROOT=.
+source %s
+kube::version::get_version_vars
+echo $KUBE_GIT_VERSION
+""" % version_script)
+        ]
+        return Subprocess(cmd, output=True).strip()
+
+    return 'unknown'
+
+
+class CIPath(object):
+    base = 'gs://kubernetes-jenkins/logs'
+    build_link = None
+    build_result_cache = None
+
+    def __init__(self, job, build):
+        self.artifacts = os.path.join(self.base, job, build, 'artifacts')
```

Comment:
I switched how this is created to hopefully make it easier to understand, take a look.